3rdparty: libpng link fails when building from source on 64bit ARM

#### System information (version)
- OpenCV => master ( 64fe0b2 )
- Operating System / Platform => ARM 64bit (Ubuntu 16.04, ODROID-C2, PINE64)
- Compiler => GCC 5.4.0
- libpng => 1.6.24 (but passes on 1.2.54)


#### Detailed description
- When trying to build OpenCV **and** libpng from source on 64bit ARM, the build fails on link stage
- Following is the summary of cmake and build log

##### log of cmake (success)
```
--     PNG:                         /usr/lib/aarch64-linux-gnu/libpng.so (ver 1.2.54)
```

##### log of cmake (failure)
```
--     PNG:                         build (ver 1.6.24)
```

##### error message of build (failure)
```
[100%] Linking CXX executable ../../bin/opencv_test_imgproc
../../lib/libopencv_imgcodecs.so.3.1.0: undefined reference to `png_init_filter_functions_neon'
collect2: error: ld returned 1 exit status
modules/imgproc/CMakeFiles/opencv_test_imgproc.dir/build.make:1245: recipe for target 'bin/opencv_test_imgproc' failed
```

- The point of this issue is ```png_init_filter_functions_neon```
- I looked a little bit in ```3rdparty/libpng``` and I think the point is NEON optimization is activated in OpenCV, but not in libpng.
- The function ```png_init_filter_functions_neon``` exists in [arm/arm_init.c](https://github.com/opencv/opencv/blob/master/3rdparty/libpng/arm/arm_init.c#L57), but it's only added to the sources(```lib_srcs```) only when ```ENABLE_NEON``` is ON as following

```cmake
if(ENABLE_NEON)
  list(APPEND lib_srcs arm/arm_init.c arm/filter_neon.S arm/filter_neon_intrinsics.c)
  add_definitions(-DPNG_ARM_NEON_OPT=2)
endif()
```

- On the other hand, this function will bw published when ```NEON``` is **available**, as following in [pngpriv.h](https://github.com/opencv/opencv/blob/master/3rdparty/libpng/pngpriv.h#L108-L140)

```c
#ifndef PNG_ARM_NEON_OPT
#  if (defined(__ARM_NEON__) || defined(__ARM_NEON)) && \
   defined(PNG_ALIGNED_MEMORY_SUPPORTED)
#     define PNG_ARM_NEON_OPT 2
#  else
#     define PNG_ARM_NEON_OPT 0
#  endif
#endif

#if PNG_ARM_NEON_OPT > 0
#  define PNG_FILTER_OPTIMIZATIONS png_init_filter_functions_neon
#endif /* PNG_ARM_NEON_OPT > 0 */
```

- Gcc on 64bit ARM always has macro ```__ARM_NEON``` defined, since NEON in part of ARMv8
- Thus, on 64bit ARM, even without ```ENEBLE_NEON=ON``` is passed to cmake, ```png_init_filter_functions_neon``` which doesn't exist, will be called.
- Usually, many Linux distribution has it's own ```libpng``` package, so the issue it self doesn't have a big impact. 
- This bug is [reported to original libpng](https://sourceforge.net/p/libpng/bugs/230/) and [fixed](https://github.com/glennrp/libpng/commit/32784523239c376dfe544bb6c7012c4934624a6d)
- I tried to fix by my own, but I wasn't good enough to fix the cmake part by my own.

#### Steps to reproduce
```
sudo apt-get -y remove libpng12-dev
mkdir build
cd build
cmake ..
make -j opencv_test_imgproc
```
Best

--- a/3rdparty/libpng/CMakeLists.txt
+++ b/3rdparty/libpng/CMakeLists.txt
@@ -31,6 +31,8 @@ file(GLOB lib_hdrs *.h)
 if(ENABLE_NEON)
   list(APPEND lib_srcs arm/arm_init.c arm/filter_neon.S arm/filter_neon_intrinsics.c)
   add_definitions(-DPNG_ARM_NEON_OPT=2)
+elseif(AARCH64)
+  add_definitions(-DPNG_ARM_NEON_OPT=0) # NEON assembler is not supported
 endif()
 
 if(ENABLE_SSE

